### PR TITLE
DNM: deleting render packages before calculating new ones creates flicker

### PR DESCRIPTION
**Pull request by mjkkirschner**
### Purpose

http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-10541

On Run Periodic there is a flash/flicker when geometry is updated.

This PR moves the request for removal of old render packages from nodes from the `cachedValue` property change event to the `OnRenderPackagesUpdated` event raised on the completion of the `UpdateRenderPackageAsyncTask `.

The reasoning is that we should have already calculated the new render packages before removing the old ones as the UI thread is doing the removal and the Scheduler thread is calculating render packages, these will be out of sync causing a flicker as the render packages are added sometime after the old ones are removed, by waiting until the new packages are calculated - removing all packages, and then adding the newly updated packages back there is no flicker.

note: I still need to run the tests on self serve CI, but have manually verified this working with run periodic and code block assignment redefinition issue. (9434).

### Declarations

Check these if you believe they are true

- [ ] The code base is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [x] Follows [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) standards (*No change of public methods or types etc..*)

### Reviewers

@ke-yu 
@aparajit-pratap 

### FYIs

@tatlin

### CODE REVIEW

**Comment by saintentropy** [comment]
LGTM....  I don't see any regressions of behavior.  In general there are some bugs in run periodic but I don't see a differences caused by this PR.  The other bugs should be filed under separate issue.

**Comment by mjkkirschner** [comment]
hmm, I see a bug with this if a node fails to execute the render package doesn't get cleaned up until it runs successfully or is deleted.

**Comment by aparajit-pratap** [comment]
@mjkkirschner I think you're changes are in the right direction. To account for the render package cleanup on node failure scenario, I suggest making the following changes:
`NodeModel.cs`
image

`DefaultWatch3DViewModel.cs`
image

**Comment by mjkkirschner** [comment]
@aparajit-pratap thanks for looking into this - I will make the changes and test it out.

**Comment by tatlin** [comment]
@mjkkirschner is this ready? We have other developers at the AAG conference interested in pushing 
realtime updates

**Comment by mjkkirschner** [comment]
@tatlin - I've made the changes suggested by @aparajit-pratap and tested them out briefly, seems to have fixed the issue I was seeing, I think this can be merged into master now.

File: src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs [hunk 1]
```diff
@@ -761,6 +760,13 @@ public override void DeleteGeometryForIdentifier(string identifier, bool request
             DeleteGeometries(geometryModels, requestUpdate); 
         }
 
+        protected override void OnRenderPackagesUpdated(NodeModel node, IEnumerable<IRenderPackage> packages)
```
Comment (aparajit-pratap) [inline src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs]: @mjkkirschner I don't think we need to override `OnRenderPackagesUpdated` here anymore as we have already made the necessary changes to it in the base class above.

Comment (mjkkirschner) [inline src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs]: If we only remove render packages if there are no updated packages then the initial code block variable reassignment issue is still present:

screen shot 2016-09-13 at 2 48 51 pm

Comment (aparajit-pratap) [inline src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs]: @mjkkirschner you're right. Not removing the render packages always will land us back in the same situation that https://github.com/DynamoDS/Dynamo/pull/6157 was trying to address in the first place. In that case you could simply change the base class method to look like this: 
```
protected virtual void OnRenderPackagesUpdated(NodeModel node, IEnumerable<IRenderPackage> packages)
        {
            
            RemoveGeometryForNode(node);
            if (packages.Any())
            {
                AddGeometryForRenderPackages(packages);
            }
        }
```

Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs]: I am fine with either approach. Moving the code to base class makes code cleaner but shuts the door completely. What if we want to customize a watch node monitoring the moving path of a point, the input keep changing and in this case we do not want to remove geometry in base class right? Maybe it's a silly idea :)

**Pull request merged**